### PR TITLE
Fix beam search to sample at least 1 non eos token (#25103)

### PR DESCRIPTION
**Pull request by yonigottesman**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by gante — COMMENTED** [review]
Thank you for the fix! 🙏 

Would you be able to:
1. Change the comment above that line to reflect the change;
2. Propagate the changes to the other beam methods as well? (`beam_sample`, `group_beam_search`, and `constrained_beam_search`)

**Comment by yonigottesman** [comment]
done

**Comment by yonigottesman** [comment]
ok now done :) 
sorry for all the force pushes I just wasn't sure how you guys merge so I preferred to keep 1 clean commit

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -3068,9 +3068,10 @@ def beam_search(
             vocab_size = next_token_scores.shape[-1]
             next_token_scores = next_token_scores.view(batch_size, num_beams * vocab_size)
 
-            # Sample 2 next tokens for each beam (so we have some spare tokens and match output of beam search)
+            # Sample 1 + len(eos_token_id) next tokens for each beam so we have at least 1 non eos tokon per beam.
```
Comment (gante) [inline src/transformers/generation/utils.py:3071]:
```suggestion
            # Sample 1 + len(eos_token_id) next tokens for each beam so we have at least 1 non eos token per beam.
```
(typo)

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -3746,9 +3747,10 @@ def group_beam_search(
                 # reshape for beam search
                 next_token_scores = next_token_scores.view(batch_size, group_size * vocab_size)
 
-                # Sample 2 next tokens for each beam (so we have some spare tokens and match output of beam search)
+                # Sample 1 + len(eos_token_id) next tokens for each beam so we have at least 1 non eos tokon per beam.
```
Comment (gante) [inline src/transformers/generation/utils.py:3750]:
```suggestion
                # Sample 1 + len(eos_token_id) next tokens for each beam so we have at least 1 non eos token per beam.
```

File: src/transformers/generation/utils.py [hunk 3]
```diff
@@ -4119,9 +4121,10 @@ def constrained_beam_search(
             vocab_size = next_token_scores.shape[-1]
             next_token_scores = next_token_scores.view(batch_size, num_beams * vocab_size)
 
-            # Sample 2 next tokens for each beam (so we have some spare tokens and match output of beam search)
+            # Sample 1 + len(eos_token_id) next tokens for each beam so we have at least 1 non eos tokon per beam.
```
Comment (gante) [inline src/transformers/generation/utils.py:4124]:
```suggestion
            # Sample 1 + len(eos_token_id) next tokens for each beam so we have at least 1 non eos token per beam.
```

**Review by gante — APPROVED** [review]
Thank you for expanding to the other beam methods as well 💪

**Comment by gante** [comment]
> sorry for all the force pushes I just wasn't sure how you guys merge so I preferred to keep 1 clean commit

@yonigottesman no worries, we squash before merging anyways ;)

**Review by sgugger — APPROVED** [review]
Thanks for your contribution! Good for me with @gante 's suggestions.

**Pull request merged**